### PR TITLE
[MCS] Awaken_invs'

### PR DESCRIPTION
**Pull request by MitchellBuckley**
This solves awaken_invs' in Schedule_R.thy

### CODE REVIEW

**Comment by mktnk3** [comment]
Nice work! I guess we have only `schedule_corres` left in this file after this.

File: proof/refine/ARM/Schedule_R.thy [hunk 1]
```diff
@@ -2847,9 +2847,19 @@ lemma setReprogramTimer_active_sc_at'[wp]:
...
 
+lemma whileM_wp_gen:
+  assumes termin:"\<And>s. I False s \<Longrightarrow> Q s"
+  assumes [wp]: "\<lbrace>I'\<rbrace> C \<lbrace>I\<rbrace>"
+  assumes [wp]: "\<lbrace>I True\<rbrace> f \<lbrace>\<lambda>_. I'\<rbrace>"
+  shows "\<lbrace>I'\<rbrace> whileM C f \<lbrace>\<lambda>_. Q\<rbrace>"
+  unfolding whileM_def
+  using termin
+  by (wpsimp wp: whileLoop_wp[where I=I])
+
```
Comment (mbrcknl) [inline proof/refine/ARM/Schedule_R.thy:2862]: I guess these rules are in some sense the wrong way around, in that `whileM_post_inv` could be proved using `whileM_wp_gen`, and `whileM_inv` could be proved using `whileM_post_inv`. No big deal, perhaps just add to the `FIXME`.
```suggestion
(* FIXME RT: move these whileM lemmas, prove `whileM_*_inv` using `whileM_wp_gen`. *)
lemmas whileM_post_inv
  = hoare_strengthen_post[where R="\<lambda>_. Q" for Q, OF whileM_inv[where P=C for C], rotated -1]

lemma whileM_wp_gen:
  assumes termin:"\<And>s. I False s \<Longrightarrow> Q s"
  assumes [wp]: "\<lbrace>I'\<rbrace> C \<lbrace>I\<rbrace>"
  assumes [wp]: "\<lbrace>I True\<rbrace> f \<lbrace>\<lambda>_. I'\<rbrace>"
  shows "\<lbrace>I'\<rbrace> whileM C f \<lbrace>\<lambda>_. Q\<rbrace>"
  unfolding whileM_def
  using termin
  by (wpsimp wp: whileLoop_wp[where I=I])

```

**Comment by mktnk3** [comment]
It looks like the merge didn’t trigger the regression test.

**Comment by lsf37** [comment]
Interesting. @LukeMondy can you have a look?

**Pull request merged**